substrate/frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,419 @@
+		) {
+			let voter_weight = Self::weight_of(Self::active_vote_of(who));
+
+			match T::Staking::status(who).expect("status checked above; qed.") {
+				StakerStatus::Nominator(nominations) => {
+					let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+						"staker should exist in VoterList, as per the contract \
+                            with staking.",
+					);
+
+					let stake_imbalance = stake_imbalance_of(prev_stake, voter_weight.into());
+
+					log!(
+						debug,
+						"on_stake_update: {:?} with {:?}. impacting nominations {:?}",
+						who,
+						stake_imbalance,
+						nominations,
+					);
+
+					// updates vote weight of nominated targets accordingly. Note: this will update
+					// the score of up to `T::MaxNominations` validators.
+					for target in nominations.into_iter() {
+						// target may be chilling due to a recent slash, verify if it is active
+						// before updating the score.
+						if <T::Staking as StakingInterface>::is_validator(&target) {
+							Self::update_score::<T::TargetList>(&target, stake_imbalance);
+						}
+					}
+				},
+				StakerStatus::Validator => {
+					// validator is both a target and a voter.
+					let stake_imbalance = stake_imbalance_of(prev_stake, voter_weight.into());
+					Self::update_score::<T::TargetList>(&who, stake_imbalance);
+
+					let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+						"the staker should exit in VoterList, as per the \
+                            contract with staking.",
+					);
+				},
+				StakerStatus::Idle => (), // nothing to see here.
+			}
+		}
+	}
+
+	// Fired when someone sets their intention to nominate.
+	//
+	// Note: it is assumed that who's staking state is updated *before* this method is called.
+	fn on_nominator_add(who: &T::AccountId) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		let _ = T::VoterList::on_insert(who.clone(), nominator_vote).defensive_proof(
+			"staker should not exist in VoterList, as per the contract with staking.",
+		);
+
+		// If who is a nominator, update the vote weight of the nominations if they exist. Note:
+		// this will update the score of up to `T::MaxNominations` validators.
+		match T::Staking::status(who).defensive() {
+			Ok(StakerStatus::Nominator(nominations)) =>
+				for t in nominations {
+					Self::update_score::<T::TargetList>(
+						&t,
+						StakeImbalance::Positive(nominator_vote.into()),
+					)
+				},
+			Ok(StakerStatus::Idle) | Ok(StakerStatus::Validator) | Err(_) => (), // nada.
+		};
+
+		log!(debug, "on_nominator_add: {:?}. role: {:?}", who, T::Staking::status(who),);
+	}
+
+	// Fired when someone sets their intention to validate.
+	//
+	// A validator is also considered a voter with self-vote and should be added to
+	// [`Config::VoterList`].
+	//
+	// Note: it is assumed that who's staking state is updated *before* calling this method.
+	fn on_validator_add(who: &T::AccountId) {
```

Comment:
Dumb question probably but, when we chill a validator, do we remove that validator from existing nominations? If not, is it not possible a validator is chilled in which case `on_validator_remove` is called and then they validate again invoking `on_validator_add` which should add all the vote weight of its existing nominators, right? In the current code, it only adds the validator's own stake as vote weight in the target list?

Reply:
That's a good point. I'm working on a couple different approaches to solve this, trying to make it elegant and without having to add more storage items, r/ws, but will share a write up on how to solve this.